include/dlaf/communication/kernels/reduce.h

```diff
@@ -37,127 +34,174 @@ namespace comm {
+           return whenAllLift(std::move(pcomm), makeItContiguous(tile)) |
+                  let_value([ex, rank_root, reduce_op](auto& pcomm, const auto& cont_buf) {
+                    return hpx::dataflow(ex, internal::reduceSend<T>, rank_root, std::move(pcomm),
+                                         reduce_op, std::cref(cont_buf));
+                  });
+         }));
+}
+}
 
-  // Note:
-  //
-  // TILE_CPU -+-> makeContiguous --> CONT_BUF --> mpi --> CONT_BUF -+-> copyBack --> CHECKPOINT -->
-  //           |                                                     |
-  //           +------------------------> TILE_CPU ------------------+----------------------------->
-  //
-  // CHECKPOINT assures that the copyBack has finished, which is useful for the next step.
+template <class T>
+void scheduleReduceRecvInPlace(const comm::Executor& ex,
+                               hpx::future<common::PromiseGuard<comm::Communicator>> pcomm,
+                               MPI_Op reduce_op, hpx::future<matrix::Tile<T, Device::CPU>> tile) {
+  using namespace hpx::execution::experimental;
 
-  auto cont_buf = dataflow(ex_copy, unwrapping(makeItContiguous_o), tile_cpu);
-  cont_buf =
-      dataflow(ex, unwrapping(reduceRecvInPlace_o), std::move(pcomm), reduce_op, std::move(cont_buf));
-  auto checkpoint = dataflow(ex_copy, unwrapExtendTiles(copyBack_o), std::move(cont_buf), tile_cpu);
+  internal::senderReduceRecvInPlace<T>(ex, std::move(pcomm), reduce_op, std::move(tile)) |
+      start_detached();
+}
 
+template <class T, Device D>
+void scheduleReduceRecvInPlace(const comm::Executor& ex,
+                               hpx::future<common::PromiseGuard<comm::Communicator>> pcomm,
+                               MPI_Op reduce_op, hpx::future<matrix::Tile<T, D>> tile) {
   // Note:
   //
-  // In case the original tile was on CPU, the previous step already managed to copy the result in place
-  // and nothing else has to be done.
-  // Otherwise, if it was on GPU, a temporary tile on CPU has been used, so this step handles the needed
-  // copy from CPU to GPU.
+  // GPU ---> Duplicate ---> (cCPU --> MPI) ---> copy --> GPU
   //
-  // TILE_CPU => no-op
-  // TILE_GPU => copy back from CPU to GPU
+  // where: cCPU = contiguous CPU
 
-  copyIfNeeded(tile_cpu, tile_orig, std::move(checkpoint));
+  using namespace hpx::execution::experimental;
+  using dlaf::internal::transform;
+
+  std::move(tile) |
+      let_value(hpx::unwrapping([ex, reduce_op, pcomm = std::move(pcomm)](auto& tile_gpu) mutable {
+        using dlaf::internal::Policy;
+        using dlaf::matrix::internal::CopyBackend;
+
+        auto tile_cpu = transform(
+            Policy<CopyBackend<D, Device::CPU>::value>(hpx::threads::thread_priority::high),
+            [](const matrix::Tile<const T, Device::GPU>& tile_gpu, auto... args) mutable {
+              return dlaf::matrix::Duplicate<Device::CPU>{}(tile_gpu, args...);
+            },
+            just(std::cref(tile_gpu)));
```

Comment:
Another option that might actually be clearer is:
```suggestion
        auto tile_cpu = transform(
            Policy<CopyBackend<D, Device::CPU>::value>(hpx::threads::thread_priority::high),
            std::bind_front(dlaf::matrix::Duplicate<Device::CPU>{}, tile_gpu));
```
I think this works and avoids the use of `reference_wrapper`. On top of that it's shorter. For the more general case (the given argument is not the first argument) there's also `bind`. `std::bind_front` is C++20 but there's `hpx::bind_front` in `hpx/local/functional.hpp`.

Alternatively, it's possible that `just` should support references... but it's quite dangerous so I'm not convinced it's a good idea to allow it.

Reply:
It's a nice suggestion, I didn't know about the existence of `bind_front`, and it's a good news it exists in HPX.

IMHO, senders are something from the standard of the future, and we are really trying to use them ahead of time thanks to HPX. We don't have concepts, which will (hopefully) simplify and improve things (e.g. compiler error messages, constraints, ...), but in this case HPX helps us by providing new features that will come, i.e. `bind_front`, and we should make use of it.

What do @teonnik and @rasolca think? Do you like @msimberg proposal?

Reply:
I agree. If some new standard features are already available in HPX feel free to use them.

Reply:
After some experiments and checks with @msimberg, we opted for not changing this code to `bind`/`bind_front`.

Main comments/notes (@msimberg feel free to correct or add details):
- the `transform` signature requires a sender: so by bounding the only argument available to the callable, a `just()` (void sender) has to be given to `transform` as 3rd parameter.
- `reference_wrappers` must be used because `just` copies the value, and `Tile` is not copyable
- Use of `reference_wrapper` introduces a template deduction problem, which requires a lambda for "supporting" the template instantiation. This because there are two "unwrapping" going on
i.e. `just(ref(tile)) -> ref(tile) -> &tile`, and the implicit conversion of `reference_wrappers` does not participate to the template deduction process, resulting (without the lambda that "guides" the template deduction) in an error (it tries to fit the `reference_wrapper` itself in the signature of the function)

Till now we haven't mentioned [`bind`](https://en.cppreference.com/w/cpp/utility/functional/bind)/[`bind_front`](https://en.cppreference.com/w/cpp/utility/functional/bind_front). Here some additional notes about them.

Let's start from reporting the main differences between `bind` and `bind_front`:
- `bind` handles in a special way `reference_wrappers`, i.e. it unwraps them calling `.get()`, while `bind_front` has no special treatment for them (apart passing them by reference, note: passing not extracting).
- `bind` requires to specify upfront the full "signature", using placeholders for not bound arguments (_If some of the arguments that are supplied in the call to g() are not matched by any placeholders stored in g, the unused arguments are evaluated and discarded_)

These differences, applied to our problem, turns out in different forms.

### std::bind

With `bind`, the blocking version gets called because without a placeholder, the additional parameter passed by `transform` will get discarded. 
```cpp
transform(
	policy,
	std::bind(dlaf::matrix::Duplicate<Device::CPU>{}, std::cref(tile_gpu)),
	just());
```

In this specific case, we could decide to "loose" generality, by adding the placeholder and so targeting just the gpu asynchronous version of `Duplicate`.
```cpp
transform(
	policy,
	std::bind(dlaf::matrix::Duplicate<Device::CPU>{}, std::cref(tile_gpu), std::placeholders::_1),
	just());
```

Note: @msimberg was wondering if the blocking call of the CUDA memcpy can be removed/hidden. What do you think? @rasolca @teonnik

### std::bind_front
About `bind_front`, we don't have the problem we had with `bind`, but we have again the one about implicit conversion (unwrapping of `reference_wrapper`) that does not participate in template deduction. So we would have to add back a lambda that guides the deduction
```cpp
transform(
	policy,
	pika::bind_front(
		[](const matrix::Tile<const T, Device::GPU>& tile_gpu, auto... args) mutable
			-> decltype(dlaf::matrix::Duplicate<Device::CPU>{}(tile_gpu, args...)) {
				return dlaf::matrix::Duplicate<Device::CPU>{}(tile_gpu, args...);
		}, std::cref(tile_gpu)),
	just());
```

Just for completeness, as just confirmed with @msimberg, it seems that the current implementation of `pika::bind_front` is not conformant with the standard (https://godbolt.org/z/nxEYxYMaT), allowing for
```cpp
transform(
	policy,
	pika::bind_front(dlaf::matrix::Duplicate<Device::CPU>{}, std::cref(tile_gpu)),
	just());
```

We won't pursue this last solution because it will not be supported in the future standard. On the other hand, as @msimberg suggested, we may think of improving `transform` in dlaf to also unwrap `reference_wrapper`s (more or less like what `bind` already does).

Reply:
Created https://github.com/eth-cscs/DLA-Future/issues/492 to keep track of possibly unwrapping `reference_wrapper` automatically.